### ISSUE TITLE
RP signout API

### ISSUE THREAD

**Issue by cbiesinger** [fedidcg/FedCM#164]
Filing this issue to discuss the shape of the RP signout API.

The spec currently has this:
https://wicg.github.io/FedCM/#browser-api-rp-sign-out

However, there are a few downsides to just calling `preventSilentAccess`:
- It is not well suited for the case of multiple client IDs on the same origin
- We don't really want to prevent silent access (the RP calling credentials.get with a silent:true argument is OK); we just don't want auto sign-in to happen

Therefore, I am proposing a different API: `FederatedCredential.prototype.logout()`.

Semantics:
- If there is no active session with the current IDP/RP/client_id pair, reject the promise
- Otherwise, remember that the user is explicitly logged out. Future attempts to autosignin will fail. (TODO: for a specific time period?)

**Comment by dj2** [comment]
Thanks for filing this, I touch on some of the changes in https://github.com/WICG/FedCM/pull/150 but that PR is currently stalled on updating some id stuff.

By `FederatedCredential.prototype.logout()` does that mean the `logout` call will happen on a returned `credential` object?

I'd probably change the semantic to always resolve the promise, need to think through if there is a privacy meaning to rejecting the promise if we decide to go that route.

**Comment by samuelgoto** [comment]
> Therefore, I am proposing a different API: FederatedCredential.prototype.logout().

If this is a method in the `prototype`, how do you get an instance of FederatedCredential? do you call `get`?

The reason I'm asking is because `get` right now, for the most part, implies a user prompt, so we either need to make it accept a `mode: silent` or something like that.

**Comment by dj2** [comment]
My assumption that if you've previously called `get` and `store` when you call `get` again it returns the stored credential (as long as preventSilentAccess has not been called). So, a second get shouldn't prompt if the user has already authenticated.

**Comment by cbiesinger** [comment]
What @dj2 said; the other option is to just create a credential manually -- `new FederatedCredential({id: "...", federated: {providers: ... }})`

**Comment by cbiesinger** [comment]
And probably return void, since it can't really fail

**Comment by samuelgoto** [comment]
> What @dj2 said; the other option is to just create a credential manually -- new FederatedCredential({id: "...", federated: {providers: ... }})

Does this throw if the consent wasn't gathered before?

> My assumption that if you've previously called get and store when you call get again it returns the stored credential (as long as preventSilentAccess has not been called). So, a second get shouldn't prompt if the user has already authenticated.

What I am asking is: would it be possible for an RP to "logout()" ONLY if the user is "logged in", but NOT prompt for a "login" in case the user isn't?

What I am imagining is that, there is a <button onclick="logout()"> that calls a logout function to log the user out. What would that function do if the user is not logged in?

Here is how I think @cbiesinger 's proposal fits in:

```
function logout() {

   const cred = await navigator.credentials.get({
       federated: { ... },
       mode: "silent" // <= this is needed, because if the user is NOT logged in, we don't WANT to log them in
   });

   if (!cred) {
     // no-op, nothing to be done here
     return;
   }

   // clears browser state
   cred.logout();
}
```

Did I get this right?

**Comment by cbiesinger** [comment]
Yes that sounds right. we can argue whether logout() should reject a promise or be a no-op if we think the user isn't logged in, I am thinking no-op.

**Comment by cbiesinger** [comment]
> Does this throw if the consent wasn't gathered before?

IMO: creating the FederatedCredential works but calling stuff on it fails

**Comment by bvandersloot-mozilla** [comment]
Would it make sense to expose the [CollectFromCredentialStore](https://w3c.github.io/webappsec-credential-management/#algorithm-collect-creds) internal method to the RP with the options specifying which IDPs to look for? Then the RP can collect `cred` whenever it wants and call its `logout` as @cbiesinger initially defined.

**Comment by samuelgoto** [comment]
> Would it make sense to expose the [CollectFromCredentialStore](https://w3c.github.io/webappsec-credential-management/#algorithm-collect-creds) internal method to the RP with the options specifying which IDPs to look for?

Yes, I think that's right. Isn't that, precisely, what [`mode=silent`](https://w3c.github.io/webappsec-credential-management/#dom-credentialmediationrequirement-silent) does?

```js
   const cred = await navigator.credentials.get({
       federated: { ... },
       mode: "silent"
   });
```

**Comment by bvandersloot-mozilla** [comment]
> > Does this throw if the consent wasn't gathered before?
> 
> IMO: creating the FederatedCredential works but calling stuff on it fails

I think that is the better of two options. That said, I prefer supporting  a `"silent"` `navigator.credentials.get` over having a functionally useless, but well formed, credential floating about and asking the dev to craft it.

**Comment by dj2** [comment]
The `logout` method has landed with the API update #194.
